codec/dagjson/marshal.go

```diff
@@ -31,21 +32,32 @@ func Marshal(n ipld.Node, sink shared.TokenSink, allowLinks bool) error {
 		if _, err := sink.Step(&tk); err != nil {
 			return err
 		}
-		// Emit map contents (and recurse).
+		// Collect map entries, then sort by key
+		type entry struct {
+			key   string
+			value ipld.Node
+		}
+		entries := []entry{}
 		for itr := n.MapIterator(); !itr.Done(); {
 			k, v, err := itr.Next()
 			if err != nil {
 				return err
 			}
-			tk.Type = tok.TString
-			tk.Str, err = k.AsString()
+			keyStr, err := k.AsString()
 			if err != nil {
 				return err
 			}
+			entries = append(entries, entry{keyStr, v})
```

Comment:
What are the implications of holding on to `v` while we collect and sort? Yeah, a perf hit, but what about memory and safety?

Reply:
The interface doesn't say whether it's safe to use the value of a Next call after the Next call has happened. Many APIs like badgerdb or Go's own `range` will reuse the same memory for the following iterations, meaning that it's on you to make a copy if you want to keep it around for longer.

I think it should be Eric's call to say what the interface semantics should be here. Allowing implementations to reuse memory would certainly unlock more performance for the cases where one doesn't need to hold onto memory, and those will likely be more common.

Reply:
(Okay.  Going to think-out-loud for a first pass here.)

Hrm.  I don't think that's actually come up as a trade before, because I think we've always been able to answer with pointers to existing and already escaped-to-heap memory.

The immutability guarantee held by ~all~ most of the node implementations so far also made it easy to say it's no problem to hold onto things.

>  Allowing implementations to reuse memory would certainly unlock more performance for the cases where one doesn't need to hold onto memory, and those will likely be more common.

So this is certainly true, we just _happen_ to not have hit a case where the distinction has been possible yet.

I can imagine in the future we might have ADL implementations which can optimize things by reusing internal buffers.  That would hit this.  I'm not immediately certain that's going to be a predominating concern in that situation, though.

Reply:
I think assuming it's safe to hold onto the values is the safer approach, anyway, as going the opposite route might well break some users.

In the future, we could always add opt-in iterator interfaces that take more aggressive shortcuts, like "may share memory", or "allows for more performant codec sorting", etc. But the basic iteration interface is... simple :)

Reply:
Yep.  Agree.  Let's go that route.

Reply:
Also worth noting here that we probably shouldn't optimise too heavily for the theoretical ADL case _here_ because this is a block codec and if you're trying to serialize a massive HAMT into DAG-{JSON,CBOR} then you're in for a bad time regardless of what optimisations might apply here. I think it's safe-ish to assume we're dealing with maps of a maximum of ~1M, maybe a little more, in the 99.9% case.

Reply:
For sorting in codecs -- yeah.  It's not particularly important to optimize for someone feeding an ADL into a marshaller.  (Even if that's technically allowed, it's... not a usual thing to want to actually do.)

The reason we ended up considering it is because it would have effect on the iterator design.

But all the same, the result here is: yeah, this is fine as written.